llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp

```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
 }
 
 // =============================================================================
-// Returns whether a `trap` intrinsic should be emitted before I.
+// Returns whether a `trap` intrinsic would be emitted before I.
 //
 // This is a copy of the logic in SelectionDAGBuilder::visitUnreachable().
 // =============================================================================
 bool NVPTXLowerUnreachable::isLoweredToTrap(const UnreachableInst &I) const {
-  if (!TrapUnreachable)
-    return false;
-  if (!NoTrapAfterNoreturn)
-    return true;
```

Comment:
> this pass would not generate traps if NoTrapAfterNoreturn was set

It seems like a trap would be generated if `NoTrapAfterNoreturn` is set and `Call->doesNotReturn()` is true. Am I missing something?

Reply:
AIUI `isLoweredToTrap` is trying to detect whether a trap would _already_ be generated by isel for any given unreachable instruction. It does this by replicating the logic for emitting traps in `SelectionDAGBuilder::visitUnreachable`. That's why this function returning true if  `NoTrapAfterNoreturn && Call->doesNotReturn()` is wrong